scripts/mempool.js

```diff
@@ -1,343 +1,242 @@
+     * `OutpointState.P2CS` will NOT return P2CS transactions.
+     * By default it's OutpointState.SPENT
+     * @param {number} [o.requirement] - A requirement to apply to all UTXOs. For example
+     * `OutpointState.P2CS` will only return P2CS transactions.
+     * By default it's MAX_SAFE_INTEGER
+     * @returns {UTXO[]} a list of unspent transaction outputs
      */
-    getUTXOs({ filter, target, onlyConfirmed = false, includeLocked }) {
-        let totFound = 0;
-        let utxos = [];
-        for (const [_, tx] of this.txmap) {
-            if (onlyConfirmed && !tx.isConfirmed()) {
+    getUTXOs({
+        filter = OutpointState.SPENT,
+        requirement = Number.MAX_SAFE_INTEGER,
+        target = Number.POSITIVE_INFINITY,
+    } = {}) {
+        const utxos = [];
+        let value = 0;
+        for (const [o, status] of this.#outpointStatus) {
+            const outpoint = COutpoint.fromUnique(o);
+            if (status & filter) {
                 continue;
             }
-            if (!this.isMature(tx)) {
+            if (!(status & requirement)) {
                 continue;
             }
-            for (let i = 0; i < tx.vout.length; i++) {
-                const vout = tx.vout[i];
-                const outpoint = new COutpoint({
-                    txid: tx.txid,
-                    n: i,
-                });
-                if (this.isSpent(outpoint)) {
-                    continue;
-                }
-                const UTXO_STATE = wallet.isMyVout(vout.script);
-                if ((UTXO_STATE & filter) == 0) {
-                    continue;
-                }
-                if (!includeLocked && wallet.isCoinLocked(outpoint)) {
-                    continue;
-                }
-                utxos.push(
-                    new UTXO({
-                        outpoint,
-                        script: vout.script,
-                        value: vout.value,
-                    })
-                );
-                // Return early if you found enough PIVs (11/10 is to make sure to pay fee)
-                totFound += vout.value;
-                if (target && totFound > (11 / 10) * target) {
-                    return utxos;
-                }
+            utxos.push(this.#outpointToUTXO(outpoint));
+            value += utxos.at(-1).value;
+            if (value >= (target * 11) / 10) {
+                break;
             }
         }
         return utxos;
     }
 
     /**
-     * Update the mempool status
-     * @param {Transaction} tx
+     * @param {number} filter
      */
-    updateMempool(tx) {
-        if (this.txmap.get(tx.txid)?.isConfirmed()) return;
-        this.txmap.set(tx.txid, tx);
-        for (const vin of tx.vin) {
-            const op = vin.outpoint;
-            if (!this.isSpent(op)) {
-                this.setSpent(op.txid, op);
-            }
-        }
-        for (const vout of tx.vout) {
-            wallet.updateHighestUsedIndex(vout);
+    getBalance(filter) {
```

Comment:
for future: this has the same running time of the old system so we can keep since this PR is just refactor...
However this operation can be optimized and we don't need to invalidate and recalculate from 0 the balance each time something changes